Ghostfury73/ArcaneScript
# Add else if support in ArcaneScript

**Description:**
This pull request addresses the issue of missing else-if functionality in ArcaneScript's grammar and its `IfElse` class implementation. The changes made enhance the script's conditional control structures to support nested conditions with else-if, improving the script's flexibility and aligning it with common programming syntax.

Closes #6 

**Changes Made:**
1. **Grammar Update:** Modified `ArcaneGrammar7.grm` to add support for else-if conditions.
   - Added new grammar rules for else-if and else-if-else scenarios.
   - New rules:
     ```plaintext
     | if '(' <Expression> ')' start <Statements> end else if '(' <Expression> ')' start <Statements> end
     | if '(' <Expression> ')' start <Statements> end else if '(' <Expression> ')' start <Statements> end else start <Statements> end
     ```

2. **IfElse.java Update:**
   - Updated the `@ProcessRule` annotation in `IfElse.java` to include new else-if grammar rules.
   - Added new fields `elseIfCondition` and `elseIfStatements` to handle else-if conditions.
   - Enhanced the `execute()` method to evaluate else-if conditions.
   - Revised the constructor to handle reduction sizes (7, 11, 15, or 19) to correctly parse and process if, if-else, if-else-if, and if-else-if-else structures.

**Code Changes Summary:**
```java
@ProcessRule(rule={
    "<Statement> ::= if ( <Expression> ) start <Statements> end",
    "<Statement> ::= if ( <Expression> ) start <Statements> end else start <Statements> end",
    "<Statement> ::= if ( <Expression> ) start <Statements> end else if ( <Expression> ) start <Statements> end",
    "<Statement> ::= if ( <Expression> ) start <Statements> end else if ( <Expression> ) start <Statements> end else start <Statements> end"
})
```

**Testing Instructions:**
1. Update `ArcaneGrammar7.grm` with the new rules.
2. Regenerate the parser if required.
3. Replace the `IfElse.java` file with the updated version provided in this PR.
4. Build the project.
5. Run the following code to validate the changes:
   ```plaintext
   a = true;
   if(a==true){
       a = false;
   } else if(a==false){
       a = true;
   }
   ```

   Expected Output:
   ```
   Execution complete!
   ```
![image](https://github.com/user-attachments/assets/20b0c09c-d9e3-45d2-9536-9f2c02fe1b31)

**Checklist:**
- [x] Grammar file updated with new else-if rules.
- [x] `IfElse.java` updated with support for else-if.
- [x] Unit tests executed and passed.
- [x] Verified output of sample test cases as expected.

---
